proposals/NNNN-Continuous-Integration-And-Testing.md

```diff
@@ -0,0 +1,190 @@
+
+Currently, the branches that constantly need code health monitoring can be classified into two types:
+
+1. **develop\master branch** - the branch that has to be under permanent testing of the main functionality. There is a potential risk that committed changes can break something. 
+2. **feature branch** - this branch type is used when new feature should be developed and added.
+
+Let's consider a few goals that have to be achieved using CI.
+
+### 1. Continuous testing on a specific branch to maintain the current working condition
+
+The following entities are specific types of checks that should be considered on the develop and feature branches:
+
+- Smoke Tests
+- Regression Tests
+- Feature Tests (applicable only for the feature branches)
+- Unit Tests + Coverage Report
+- Coding Style Check
+- Static Code Analysis
+- Dynamic Analysis on Smoke Tests
+
+**! Note that each of the above checks (except Coding Style Check and Static Code Analysis) should be run for all supported Policies flows (External proprietary, Proprietary, HTTP).**
+
+Also, it would be nice to perform project compilation with the different sets of flags: with and without ENABLE_LOG, ENABLE_SECURITY, DEBUG/RELEASE mode etc - since this allows to find hidden problems.
+
+#### ATF/smoke tests running
+
+There are smoke tests to check the main life functions of SDL (such as `run`, `shutdown`, etc.) This type of tests should be run for each `pull` and `push` events. Smoke testing, also known as build acceptance testing (BAT), is a critical aspect of quality assurance that delivers quick and decisive answers on the viability of a particular build.
+
+Ideally, each smoke test should meet these criteria:
+
+- Tests core features only
+- The test should be repeatable indefinitely
+- Very fast execution
+- Should generate few or no false positives
+
+Now SDL includes the following scopes for smoke testing: Registration, API, Heartbeat, Policy, Shutdown, Resumption.
+
+#### ATF/regression tests running
+
+Regression Testing is a full or partial selection of already executed test cases which are re-executed to ensure existing functionality works fine. This testing is done to make sure that new code changes should not have side effects on the existing functionality. It ensures that the old code still works once the new code changes are done. When the new feature is accepted and corresponding code is integrated to sdl_core a new corresponding test scope is integrated to the sdl_atf_test_scripts as well as regression tests extended.
+
+#### Feature Tests
+
+The aim of feature testing is to make sure that new changes for SDL works properly and meet all the intended specifications. All bugs and defects in the features should be found during testing stages otherwise it could cause huge productivity loss. By performing feature testing CI can improve quality of the end product and deliver products with exceptional functionality.
```

Comment:
```suggestion
The aim of feature testing is to make sure that new changes to SDL Core work properly and meet all the intended specifications. All bugs and defects in the features should be found during testing stages otherwise it could cause huge productivity loss. By performing feature testing, CI can improve quality of the end product and deliver products with exceptional functionality.
```